### ISSUE TITLE
Istio CNI Install not using current ServiceAccount tokens

### ISSUE THREAD

**Issue by tommyp1ckles** [istio/istio#38077]
### Bug Description

On k8s clusters running 1.21+ mounted service account tokens can be created with a expiry. I've started noticing nodes getting into states with errors like `failed to set up sandbox container ... network: Unauthorized`.

This appears to be due to expired or invalid tokens being used in `/etc/cni/net.d/ZZZ-istio-cni-kubeconfig`. Initially we thought this was an aws-vpc-cni issue since restarting that pod would fix the issue on that node.

However this appears to only fix the problem since this triggers a filesystem notify event here: https://github.com/istio/istio/blob/1.11.4/cni/pkg/install/install.go#L78 which causes a full configuration reload, including the kubeconfig file.

Should this loop also not be watching for updates to the mounted service account token as well?

### Version

```prose
Istio Version: 1.11.1
Kubernetes Version: v1.21.5-eks-bc4871b
```

(edit: that was k8s 1.21, not 1.12).

**Comment by jbilliau-rcd** [comment]
Im having this exact same issue on multiple clusters, probably a dozen times over the past 3 months. Running Istio 1.11.4 and EKS 1.21. Thought it was aws-node related, but there are no errors or anything, and then seeing that issue now makes me think it's Istio related?

**Comment by jbilliau-rcd** [comment]
It looks like istio-cni needs to be updated to support `BoundServiceAccountTokens`. See email we just got from AWS:

> Kubernetes version 1.21 graduated BoundServiceAccountTokenVolume feature [1] to beta and enabled it by default. This feature improves security of service account tokens by requiring a one hour expiry time, over the previous default of no expiration. This means that applications that do not refetch service account tokens periodically will receive an HTTP 401 unauthorized error response on requests to Kubernetes API server with expired tokens. You can learn more about the BoundServiceAccountToken feature in EKS Kubernetes 1.21 release notes [2].
> 
> To enable a smooth migration of applications to the newer time-bound service account tokens, EKS v1.21+ extends the lifetime of service account tokens to 90 days. Applications on EKS v1.21+ clusters that make API server requests with tokens that are older than 90 days will receive an HTTP 401 unauthorized  error response.
> 
> How can you resolve the issue?
> 
> To make the transition to time bound service account tokens easier, Kubernetes has updated the below official versions of client SDKs to automatically refetch tokens before the one hour expiration:
> 
> * Go v0.15.7 and later
> * Python v12.0.0 and later
> * Java v9.0.0 and later
> * Javascript v0.10.3 and later
> * Ruby master branch
> * Haskell v0.3.0.0
> 
> We recommend that you update your application and its dependencies to use one of the above client SDK versions if you are on an older version. 
> 
> While not an exhaustive list, the below AWS components have been updated to use the newer Kubernetes client SDKs that automatically refetches the token :
> 
> * Amazon VPC CNI: v1.8.0 and later
> * CoreDNS: v1.8.4 and later
> * AWS Load Balancer Controller: v2.0.0 and later
> * kube-proxy: v1.21.2-eksbuild.2 and later 
> 
> You can find instructions for upgrading Amazon VPC CNI, CoreDNS, AWS Load Balancer Controller and kube-proxy in EKS documentation at [3], [4], [5] and [6] respectively. Please check EKS documentation [2] for list of components and the versions that have addressed this issue. 
> 
> How can I identify service accounts using stale tokens?
> 
> As of April 20th 2022, we have identified the below service accounts attached to pods in one or more of your EKS clusters using stale (older than 1 hour) tokens. Service accounts are listed in the format : <eks-cluster-arn>|<namespace:serviceaccount>
> 
> **arn:aws:eks:us-west-2:999999999:cluster/processing-prod-999999999-us-west-2|kube-system:istio-cni**
> 
> 
> You can also use the CloudWatch Log insights query provided in the troubleshooting section of EKS documentation [7] to identify the current list of pods using service accounts with stale tokens and the time elapsed since the token was created. 
> 
> We recommend that you update your applications and its dependencies that are using stale service accounts tokens to use one of the newer Kubernetes Client SDKs that refetches tokens. 
> 
> If the service account token used is close to expiry (<90 days) and you do not have sufficient time to update your client SDK versions before expiry, then you can terminate existing pods and create new ones. This results in refetching of the service account token, giving you additional time (90 days) to update your client SDKs._
> 
>

**Comment by pavelpanfilov** [comment]
got same email from AWS, EKS v1.22 istio v1.12.4

**Comment by malpania** [comment]
We did temporary workaround to solve the istio cni reload issue. 
We are running a demonset pod and running the command. 
```
  command: ["/bin/sh", "-c", "while true; do sleep 3600 & wait ; echo \"$(date -u '+%Y-%m-%dT%H:%M:%SZ') [INFO] Forcing Istio CNI conf install refresh\" ; rm -f /host/etc/cni/net.d/istio-cni-reloader.txt ; printf 'this file exists to reload kubeconfig for istio-cni' >> /host/etc/cni/net.d/istio-cni-reloader.txt ; done"]
```

**Comment by jbilliau-rcd** [comment]
@malpania ah good info, thanks for the suggestion. We ended up already "batch solving" this by making a CronJob that redeploys aws-node, as well as 3 other controllers, all with the same problem. We are going to keep this going until all the various controllers as well as Istio itself fixes this issue.

**Comment by haani-niyaz** [comment]
We encountered the same issue. Pod restarts did not fix the issue. It was resolved when we upgraded istio to 1.13.4 which in turn upgraded the istio-cni.

**Comment by guiguan** [comment]
We are on istio-cni 1.14.0 now, and we are still observing the stale service token problem

If we query our control plane logs within last hour in CloudWatch as [suggested by AWS](https://docs.aws.amazon.com/eks/latest/userguide/service-accounts.html#service-account-tokens), both `kiali` and `istio-cni` are still in the results

<img width="791" alt="Screen Shot 2022-06-15 at 4 57 36 pm" src="https://user-images.githubusercontent.com/940189/173767773-2c262055-b26b-4ec5-bdfe-78132cf762ba.png">

**Comment by studees** [comment]
I am facing this also on Istio 1.14 - please can someone respond to this ticket to confirm workaround if you believe it is not a real issue?

**Comment by tabnul** [comment]
when will this issue be solved / merged in the codebase? for us its a blocker to go live with istio. (currently in project phase)

what could be a workaround?>

**Comment by joze144** [comment]
@tabnul 

My workaround is a script that restarts istio-cni-node DaemonSet every 2 months or so.

`$ kubectl rollout restart ds/istio-cni-node -n kube-system`

That seems to solve the issue for me, but it's annoying.

**Comment by georgiiturevich-uw** [comment]
@tabnul The fix was merged to master here https://github.com/istio/istio/commit/6bbb8519b98a578cbbed657916417b7a440962db and was released together with Istio 1.15
